typing/rec_check.ml

```diff
@@ -18,145 +18,161 @@ open Types
-      let r =
-      Ident.fold_all
-        (fun id v tbl ->
-           let v' = try Ident.find_same id tbl with Not_found -> Use.empty in
-           Ident.add id (Use.join v v') tbl)
-        x
-        y
-      in
-      r
-  end
+module Mode = struct
+  (** For an expression in a program, its "usage mode" represents
+     static information about how the value produced by the expression
+     will be used by the context around it. *)
+  type t =
+    | Dereference
+    (** A [Dereference] context consumes, inspects and uses the value
+        in arbitrary ways. Such a value must be fully defined at the point
+        of usage, it cannot be defined mutually-recursively with its context. *)
+
+    | Return
+    (** A [Return] context returns its value without further inspection.
+        This value cannot be defined mutually-recursively with its context,
+        as there is a risk of self-loop: in [let rec x = y and y = x], the
+        two definitions use a single variable in [Return] context. *)
+
+    | Guard
+    (** A [Guard] context returns the value as a member of a data structure,
+        for example a variant constructor or record. The value can safely be
+        defined mutually-recursively with their context, for example in
+        [let rec li = 1 :: li].
+        When these subexpressions participate in a cyclic definition,
+        this definition is productive/guarded.
+
+        The [Guard] mode is also used when a value is not dereferenced,
+        it is returned by a sub-expression, but the result of this
+        sub-expression is discarded instead of being returned.
+        For example, the subterm [?] is in a [Guard] context
+        in [let _ = ? in e] and in [?; e].
+        When these subexpressions participate in a cyclic definition,
+        they cannot create a self-loop.
+    *)
+
+    | Delay
+    (** A [Delay] context can be fully evaluated without evaluting its argument,
+        which will only be needed at a later point of program execution. For
+        example, [fun x -> ?] or [lazy ?] are [Delay] contexts. *)
+
+    | Unused
+    (** [Unused] is for subexpressions that are not used at all during
+       the evaluation of the whole program. This is the mode of
+       a variable in an expression in which it does not occur. *)
+
+  (* Returns the most conservative mode of the two arguments.
+
+     Dereference < Return < Guard < Delay < Unused
+
+     In judgments we write (m + m') for (join m m').
+  *)
+  let join m m' =
+    match m, m' with
+    | Dereference, _
+    | _, Dereference -> Dereference
+    | Return, _
+    | _, Return -> Return
+    | Guard, _
+    | _, Guard -> Guard
+    | Delay, _
+    | _, Delay -> Delay
+    | _ -> Unused
+
+  (* If x is used with the mode m in e[x], and e[x] is used with mode
+     m' in e'[e[x]], then x is used with mode m'[m] in e'[e[x]].
+
+     Return is neutral for composition: m[Return] = m = Return[m].
+  *)
+  let compos m' m = match m', m with
```

Comment:
Minor: `compos`?  Why not `compose`?

Aside: 'not' patterns would be useful here.

Besides noting that `Return` is neutral for composition, you might also note that `compos` is associative (I checked that it is), and that `Unused` is an annihilator.